src/erc721.rs

```diff
@@ -0,0 +1,367 @@
+        //     let sender = ensure_signed(origin)?;
+        //     ensure!(Self::_is_approved_or_owner(sender, token_id), "You do not own this token");
+
+        //     Self::_transfer_from(from, to, token_id)?;
+
+        //     Ok(())
+        // }
+
+        // safe_transfer_from checks that the recieving address has enough balance to satisfy the ExistentialDeposit
+        // This is not quite what it does on Ethereum, but in the same spirit...
+        // fn safe_transfer_from(origin, from: T::AccountId, to: T::AccountId, token_id: T::Hash) -> DispatchResult {
+        //     let to_balance = <balances::Module<T>>::free_balance(&to);
+        //     ensure!(!to_balance.is_zero(), "'to' account does not satisfy the `ExistentialDeposit` requirement");
+
+        //     Self::transfer_from(origin, from, to, token_id)?;
+
+        //     Ok(())
+        // }
+    }
+}
+
+impl<T: Trait> Module<T> {
+    // Create a new token
+    pub(crate) fn create_token(
+        account_id: &T::AccountId,
+    ) -> sp_std::result::Result<T::Hash, DispatchError> {
+        let nonce = Nonce::get();
+
+        // generate random hash based on account id and nonce
+        let random_hash = (
+            <T as Trait>::Randomness::random_seed(),
+            account_id.clone(),
+            nonce,
+        )
+            .using_encoded(<T as system::Trait>::Hashing::hash);
+
+        Self::_mint(account_id, &random_hash.into())?;
+        Nonce::mutate(|n| *n += 1);
+
+        Ok(random_hash)
+    }
+
+    // If token is existed
+    pub(crate) fn _exists(token_id: &T::Hash) -> bool {
+        return <TokenOwner<T>>::exists(token_id);
+    }
+
+    // Token owner or token approval or owner's delegate
+    pub(crate) fn _is_approved_or_owner(spender: &T::AccountId, token_id: &T::Hash) -> bool {
+        let owner = Self::owner_of(token_id);
+        let approved_user = Self::get_approved(token_id);
+
+        let approved_as_owner = match owner {
+            Some(ref o) => o == spender,
+            None => false,
+        };
+
+        let approved_as_delegate = match owner {
+            Some(d) => Self::is_approved_for_all((d, spender.clone())),
+            None => false,
+        };
+
+        let approved_as_user = match approved_user {
+            Some(u) => u == *spender,
+            None => false,
+        };
+
+        return approved_as_owner || approved_as_user || approved_as_delegate;
+    }
+
+    // Mint a new token
+    pub(crate) fn _mint(to: &T::AccountId, token_id: &T::Hash) -> DispatchResult {
+        ensure!(!Self::_exists(token_id), "Token already exists");
+
+        let balance_of = Self::balance_of(to);
+
+        let new_balance_of = match balance_of.checked_add(1) {
+            Some(c) => c,
+            None => return Err(Error::<T>::Overflow.into()),
+        };
+
+        Self::_add_token_to_all_tokens_enumeration(token_id)?;
+        Self::_add_token_to_owner_enumeration(to, token_id)?;
```

Comment:
1 the potential error is from number overflow or underflow. Actually, overflow almost impossible since we give a u64 integer. And underflow could be avoid if we always put a increment before a decrement. I have noticed some pallet neglect such error. Just do increment/decrement operations.
2. I think the real state change will be applied after the whole extrinsic successfully executed in WASM EM. I am not sure it will get a broken state. I will try to verify in real runtime.

Reply:
Agree that it is unlikely, but we should still make sure that the state cannot be inconsistent. Please check for underflow/overflow first, and write afterwards. 

I'd be also fine to use saturating methods (not throwing, but rather setting to 0 if underflow happens, or to 2^64-1 if overflow happens).